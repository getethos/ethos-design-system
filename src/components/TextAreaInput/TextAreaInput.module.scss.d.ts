// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'Resize': string;
  'TextAreaInput': string;
}
export const cssExports: CssExports;
export default cssExports;
