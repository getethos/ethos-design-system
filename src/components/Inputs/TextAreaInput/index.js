export { TextAreaInput } from './TextAreaInput.js'
