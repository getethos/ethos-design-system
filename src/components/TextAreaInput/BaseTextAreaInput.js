import React from 'react'
import PropTypes from 'prop-types'

export const BaseTextAreaInput = ({
  className,
  disabled,
  name,
  placeholder,
  onPaste,
  onBlur,
  onFocus,
  onChange,
  value,
  ...rest
}) => {
  return (
    <textarea
      className={className}
      disabled={disabled}
      name={name}
      placeholder={placeholder}
      onPaste={onPaste}
      onChange={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
      value={value}
      data-tid={rest['data-tid']}
    />
  )
}

BaseTextAreaInput.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  onPaste: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  onChange: PropTypes.func,
  value: PropTypes.string,
  'data-tid': PropTypes.string.isRequired,
}
