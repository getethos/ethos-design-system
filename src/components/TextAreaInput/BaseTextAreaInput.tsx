import * as React from 'react'
import { PublicFormProps } from '../Form/Form'
type BaseTextAreaInputProps = PublicFormProps & {
  className?: string
  disabled?: boolean
  placeholder?: string
  onPaste?: (...args: any[]) => any
  onBlur?: (...args: any[]) => any
  onFocus?: (...args: any[]) => any
  onChange?: (...args: any[]) => any
  value?: string
}
/**
 * @private
 *
 * Required props:
 * @param {string} name
 * @param {string} 'data-tid'
 */
export const BaseTextAreaInput: React.FC<BaseTextAreaInputProps> = ({
  className,
  disabled,
  name,
  placeholder,
  onPaste,
  onBlur,
  onFocus,
  onChange,
  value,
  ...rest
}) => {
  return (
    <textarea
      className={className}
      disabled={disabled}
      name={name}
      placeholder={placeholder}
      onPaste={onPaste}
      onChange={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
      value={value}
      data-tid={rest['data-tid']}
    />
  )
}
