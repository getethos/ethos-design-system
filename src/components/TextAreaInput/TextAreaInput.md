### Default

```jsx
// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}

;<TextAreaInput
  name="example"
  labelCopy="TextAreaInput"
  data-tid="the-textarea-input"
  formChangeHandler={formChangeHandlerStub}
  validator={() => 'This always fails'}
/>
```

### Handles resize prop, only vertical because it looks nicer

```jsx
;<TextAreaInput
  resize={true}
  name="example"
  labelCopy="Resize me"
  data-tid="the-textarea-input"
/>
```
