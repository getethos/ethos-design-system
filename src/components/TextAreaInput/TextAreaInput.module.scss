@import '../TextInput/TextInput.module.scss';

.TextAreaInput {
  composes: TextInputCommon TextInputStylable;
  resize: none;
}
.Resize {
  resize: vertical;
}
