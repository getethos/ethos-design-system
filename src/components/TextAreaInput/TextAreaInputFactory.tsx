import * as React from 'react'
import { PrivateTextAreaInput } from './PrivateTextAreaInput'
export function TextAreaInputFactory(privateProps) {
  const PublicTextAreaInputComponent = (downstreamProps) => {
    return <PrivateTextAreaInput {...downstreamProps} {...privateProps} />
  }
  return PublicTextAreaInputComponent
}
