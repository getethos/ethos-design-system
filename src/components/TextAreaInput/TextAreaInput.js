import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import { InputLabel } from '../Inputs/InputLabel'
import useRequired from '../../hooks/useRequired.js'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import useInvalid from '../../hooks/useInvalid.js'
import useInputValidation from '../../hooks/useInputValidation.js'
import restrict from '../../helpers/restrict.js'
import styles from './TextAreaInput.module.scss'
import errorStyles from '../Inputs/Errors.module.scss'

/**
 * @param  {String}   props.name        Input name and htmlFor prop for label
 * @param  {String}   props.labelCopy   User-visible text of label for input
 * @param  {Boolean}  props.allCaps     Whether to text-trasform: uppercase
 * @param  {Function} props.validator   Function for validating input
 * @param  {Boolean}  props.disabled
 */

function PrivateTextAreaInput({
  resize,
  disabled,
  name,
  labelCopy,
  allCaps,
  formChangeHandler,
  validator,
  initialValue,
  currentValue,
  currentError,
  setFieldTouched,
  restrictIllegal,
  ...rest
}) {
  // Verify that all required props were supplied
  const [includesRequired] = useRequired(['data-tid', 'name', 'labelCopy'])
  let allRelevantProps = Object.assign({}, rest, {
    name: name,
    labelCopy: labelCopy,
    allCaps: allCaps,
  })
  includesRequired(allRelevantProps)

  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(
    Object.keys(PrivateTextAreaInput.PUBLIC_PROPS)
  )
  includesInvalid(rest)

  // Set up validation hooks
  const [getError, setError, , validate] = useErrorMessage(validator)

  const [value, setValue] = useState(currentValue || initialValue || '')

  const [touched, setTouched] = useState(initialValue ? true : false)

  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
  })

  const onChange = (ev) => {
    const val = ev.target.value
    const restrictedVal = restrictIllegal ? restrict(val) : val
    setValue(restrictedVal)

    // We call setTouched in onBlur, so can reliably call getter here
    doValidation(restrictedVal, touched)
  }

  const setAllTouched = () => {
    // We set touched to change the react state, but it's async and
    // processing still, so, we use a flag for doValidation
    setTouched(true)
    // Also tell the form we've been touched
    if (setFieldTouched) {
      setFieldTouched(true)
    }
  }

  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      console.log('TextAreaInput useEffect -- calling doValidation')
      doValidation(initialValue, '')
    }
  }, [])

  const onBlur = (ev) => {
    setAllTouched()
    doValidation(ev.target.value, true)
  }

  const onPaste = (ev) => {
    const val = ev.clipboardData.getData('text/plain')
    const restrictedVal = restrictIllegal ? restrict(val) : val
    setValue(restrictedVal)
  }

  const classes = [styles.TextAreaInput]
  if (getError(currentError, touched)) {
    classes.push(errorStyles.Error)
  }
  if (resize) {
    classes.push(styles.Resize)
  }

  return (
    <>
      <InputLabel name={name} labelCopy={labelCopy} />
      <textarea
        className={classes.join(' ')}
        disabled={disabled}
        name={name}
        placeholder={rest.placeholder}
        onPaste={onPaste}
        onChange={onChange}
        onBlur={onBlur}
        value={value}
        data-tid={rest['data-tid']}
      />
      {getError(currentError, touched)}
    </>
  )
}

PrivateTextAreaInput.PUBLIC_PROPS = {
  resize: PropTypes.bool,
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  allCaps: PropTypes.bool,
  initialValue: PropTypes.string,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
  placeholder: PropTypes.string,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  restrictIllegal: PropTypes.bool,
}

PrivateTextAreaInput.propTypes = {
  ...PrivateTextAreaInput.PUBLIC_PROPS,
}

PrivateTextAreaInput.defaultProps = {
  placeholder: '',
  restrictIllegal: true,
}

function TextAreaInputFactory(privateProps) {
  const PublicTextAreaInputComponent = (downstreamProps) => {
    return <PrivateTextAreaInput {...downstreamProps} {...privateProps} />
  }

  return PublicTextAreaInputComponent
}

export const TextAreaInput = TextAreaInputFactory()
