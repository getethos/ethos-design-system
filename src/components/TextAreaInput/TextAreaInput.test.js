import React from 'react'
import { TextAreaInput } from './TextAreaInput'
import renderer from 'react-test-renderer'

describe('TextAreaInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <TextAreaInput
          name="nombre"
          resize={true}
          allCaps={true}
          labelCopy="The label"
          data-tid="the-text-input"
          validator={(x) =>
            x.length % 2
              ? 'Text does not have an even number of characters'
              : ''
          }
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
