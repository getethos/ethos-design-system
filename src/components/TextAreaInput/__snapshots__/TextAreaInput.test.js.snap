// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaInput default rendering 1`] = `
Array [
  <label
    className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
    htmlFor="nombre"
  >
    The label
  </label>,
  <div
    aria-hidden={true}
    className="Spacer"
    style={
      Object {
        "height": 4,
      }
    }
  />,
  <textarea
    className="TextAreaInput Resize"
    data-tid="the-text-input"
    name="nombre"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    value=""
  />,
  "",
]
`;
